File: LEETCODE/LEETCODE-CPP/78.cpp
//78. Subsets
//https://leetcode.com/problems/subsets/description/

//Method-1 using backtracking & recursion
class Solution {
    public:
        vector<vector<int>> subsets(vector<int>& nums) {
            vector<vector<int>> ans;
            vector<int> temp;
    
            helper(nums,ans,temp,0);
    
            return ans;
        }
    
        void helper(vector<int>&nums,vector<vector<int>>&ans,vector<int>temp, int idx){
            ans.push_back(temp);
    
            for(int i=idx;i<nums.size();i++){
                temp.push_back(nums[i]);
                helper(nums,ans,temp,i+1);
                temp.pop_back();
            }
        }
    };

File: LEETCODE/LEETCODE-JAVA/78.java
//78. Subsets
//https://leetcode.com/problems/subsets/description/

//Method-1 using backtracking & recursion
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        int n = nums.length;

        helper(nums, ans, temp, n, 0);
        return ans;
    }

    private void helper(int[] nums, List<List<Integer>> ans, List<Integer> temp, int n, int idx) {
        ans.add(new ArrayList<>(temp));  

        for (int i = idx; i < n; i++) {
            temp.add(nums[i]);                      
            helper(nums, ans, temp, n, i + 1);       
            temp.remove(temp.size() - 1);            
        }
    }
}